rfcs/45-stageless.md

```diff
@@ -0,0 +1,866 @@
+- system sets (use `.add_systems` instead)
+- stages (yoten)
+
+### Scheduling overview and intent-based configuration
+
+Systems in Bevy are stored in a `Schedule`: a collection of **configured systems**.
+In each iteration of the schedule (typically corresponding to a single frame), the `App`'s `runner` function will run the schedule,
+causing the **scheduler** to run the systems in parallel using a strategy that respects all of the configured constraints.
+If these constraints cannot be met (for example, a system may want to run both before and after another system), the schedule is said to be **unsatisfiable**, and the scheduler will panic.
+Schedules can be nested and branched from within **exclusive systems** (which have mutable access to the entire `World`), allowing you to encode arbitrarily complex control flow using ordinary Rust constructs.
+
+By default, each new `Schedule` is entirely unordered: systems will be selected in an arbitrary order (which is not guaranteed to be stable across passes of the schedule) and run if and only  if all of the data that they must access are free.
+Just like with standard borrow checking, multiple systems can read from the same data at once, but writing to the data requires an exclusive lock.
+Systems which cannot be run in parallel are said to be **incompatible**.
+
+In Bevy, schedules are ordered via **intent-based configuration**: record the exact, minimal set of constraints needed for a scheduling strategy to be valid, and then let the scheduler figure it out!
+As the number of systems in your app grows, the number of possible scheduling strategies grows exponentially, and the set of *valid* scheduling strategies grows almost as quickly.
+However, each system will continue to work hand-in-hand with a small number of closely-related systems, interlocking in a straightforward and comprehensible way.
+
+By configuring our systems with straightforward, local rules, we can allow the scheduler to pick one of the possible valid paths, optimizing as it sees fit without breaking our logic.
+Just as importantly, we are not over-constraining our ordering. Subtle ordering bugs will surface quickly and then be stamped out with a well-considered rule.
+
+### The `App` stores multiple schedules
+
+The `App` can store multiple `Schedules`: each with their own `impl ScheduleLabel` type, allowing you to cleanly execute coherent blocks of logic when the need arises.
+By default, each app stores both a startup and main schedule: the former runs only once on app creation, while the latter loops endlessly.
+However, schedules are a surprisingly powerful and flexible tool: they can be used to handle initialization and cleanup logic when working with states or used ad-hoc to run game logic in complex patterns.
+
+The main, startup and rendering schedules can be accessed using the `DefaultSchedule::Main`, `DefaultSchedule::Startup`, `DefaultSchedule::Rendering` labels respectively.

```

Comment:
We might need a fourth `DefaultSchedule` for a place to insert systems that can modify Main Schedule.

Reply:
Since any schedule could generate schedule commands to edit `DefaultSchedule::Main`, why?

Reply:
My feeling is that schedule modification should just *always* be deferred.

Reply:
My concern was that we might not be able to read the `Main` schedule while running it. i.e. that we'll need exclusive access to the schedule to run it.  If we can avoid that then we probably won't need an empty schedule.

Reply:
> My concern was that we might not be able to read the Main schedule while running it

This should be safe: any mutated information should almost certainly not be publically exposed.